src/extension.rs

```diff
@@ -0,0 +1,881 @@
+    /// This works ok because dyn ExtensionValues should never be seen in user code.
+    /// See `ExtensionSet`.
+    fn inner_eq(&self, other: &dyn ExtensionValue) -> bool;
+
+    /// True if the inner type is considered to have data.
+    fn has_data(&self) -> bool;
+
+    /// ProtoIntType of the underlying type, cached for encoding.
+    fn proto_int_type(&self) -> ProtoIntType;
+}
+
+/// A concrete holder for strongly typed extension data.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct ExtensionValueImpl<T> {
+    // This is an Option so the data can be moved out from a &mut.
+    data: Option<T>,
+
+    proto_int_type: ProtoIntType,
+}
+
+impl<T> ExtensionValueImpl<T> {
+    pub fn data(&self) -> Option<&T> {
+        self.data.as_ref()
+    }
+    pub fn data_mut(&mut self) -> Option<&mut T> {
+        self.data.as_mut()
+    }
+    pub fn set_data(&mut self, value: T) {
+        self.data = Some(value);
+    }
+    pub fn take_data(&mut self) -> Option<T> {
+        self.data.take()
+    }
+}
+
+impl<T> Merge for ExtensionValueImpl<T>
+where
+    T: Merge + Default,
+{
+    fn merge(
+        &mut self,
+        proto_int_type: ProtoIntType,
+        wire_type: WireType,
+        buf: &mut MergeBuffer,
+        ctx: DecodeContext,
+    ) -> Result<(), DecodeError> {
+        let data = self.data.get_or_insert_with(T::default);
+        data.merge(proto_int_type, wire_type, buf, ctx)
+    }
+}
+
+impl<T> Encode for ExtensionValueImpl<T>
+where
+    T: Encode,
+{
+    fn encode(&self, proto_int_type: ProtoIntType, tag: u32, buf: &mut EncodeBuffer) {
+        if let Some(data) = &self.data {
+            data.encode(proto_int_type, tag, buf);
+        }
+    }
+
+    fn encoded_len(&self, proto_int_type: ProtoIntType, tag: u32) -> usize {
+        match &self.data {
+            None => 0,
+            Some(data) => data.encoded_len(proto_int_type, tag),
+        }
+    }
+}
+
+impl<T> ExtensionValue for ExtensionValueImpl<T>
+where
+    T: 'static + Merge + Encode + Clone + PartialEq + Default + Debug,
+{
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn inner_clone(&self) -> Box<dyn ExtensionValue> {
+        let cloned = (*self).clone();
+        Box::new(cloned)
+    }
+
+    fn inner_eq(&self, other: &dyn ExtensionValue) -> bool {
+        let any = other.as_any().downcast_ref::<ExtensionValueImpl<T>>();
+        match any {
+            None => false,
+            Some(other) => self == other,
+        }
+    }
+
+    fn has_data(&self) -> bool {
+        self.data.is_some()
+    }
+
+    fn proto_int_type(&self) -> ProtoIntType {
+        self.proto_int_type
+    }
+}
+
+/// A generic container for an Extension that can delegate to the implementation for type-specific work.
+pub trait Extension {
+    /// Fully-qualified type name of the message this extension is for.
+    fn extendable_type_id(&self) -> ExtendableTypeId;
```

Comment:
This also could return a borrowed string to allow implementing for something like [`ExtensionDescriptor`](https://docs.rs/prost-reflect/0.8.1/prost_reflect/struct.ExtensionDescriptor.html)